rust/src/domains/polars/expr/mod.rs

```diff
@@ -0,0 +1,214 @@
+use polars::lazy::dsl::Expr;
+use polars::prelude::*;
+use std::collections::BTreeSet;
+use std::fmt::{Debug, Formatter};
+
+use crate::core::{Metric, MetricSpace};
+use crate::metrics::{
+    AbsoluteDistance, ChangeOneDistance, HammingDistance, InsertDeleteDistance, LInfDistance,
+    LpDistance, PartitionDistance, SymmetricDistance,
+};
+use crate::traits::ProductOrd;
+use crate::transformations::DatasetMetric;
+use crate::{core::Domain, error::Fallible};
+
+use super::{Frame, FrameDomain, LogicalPlanDomain, NumericDataType, SeriesDomain};
+
+#[cfg(feature = "ffi")]
+mod ffi;
+
+// TODO: remove this allow marker later
+#[allow(dead_code)]
+#[derive(Clone, PartialEq, Debug)]
+pub enum ExprContext {
+    // select
+    Select,
+    // filter
+    Filter,
+    // with_column
+    WithColumns,
+    // group_by
+    Aggregate(BTreeSet<String>),
+}
+
+impl ExprContext {
+    fn get_plan(&self, val: &(LogicalPlan, Expr)) -> LogicalPlan {
+        let (lp, expr) = val.clone();
+        let frame = LazyFrame::from(lp);
+        match self {
+            ExprContext::Select => frame.select([expr]),
+            ExprContext::Filter => frame.filter(expr),
+            ExprContext::WithColumns => frame.with_columns([expr]),
+            ExprContext::Aggregate(columns) => frame
+                .group_by(&columns.iter().map(|n| col(n)).collect::<Vec<_>>())
+                .agg([expr]),
+        }
+        .logical_plan
+    }
+
+    pub fn break_alignment(&self) -> Fallible<()> {
+        if matches!(self, ExprContext::Filter | ExprContext::WithColumns) {
+            return fallible!(
+                MakeMeasurement,
+                "record alignment can only be broken in a selection or aggregation"
+            );
+        }
+        Ok(())
+    }
+
+    pub fn grouping_columns(&self) -> BTreeSet<String> {
+        match self {
+            ExprContext::Aggregate(columns) => columns.clone(),
+            _ => BTreeSet::new(),
+        }
+    }
+}
+
+/// # Proof Definition
+/// `ExprDomain` is the domain of all polars expressions that can be applied to a `LazyFrame` where:
+/// * `lazy_frame_domain` - `LazyFrameDomain`.
+/// * `context` - Context in which expression is to be applied.
+/// * `active_column` - Column to which expression is to be applied.
```

Comment:
Should this correspond to the code in the struct? It has `frame_domain` instead of `lazy_frame_domain`, and doesn't mention `active_column`.